cmd/contour/shutdownmanager.go

```diff
@@ -0,0 +1,132 @@
+// Copyright © 2020 VMware
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+	"time"
+
+	"github.com/projectcontour/contour/internal/metrics"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcontour/contour/internal/workgroup"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+// handler for /healthz
+func (s *shutdownmanagerContext) healthzHandler(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+	_, err := w.Write([]byte("OK"))
+	if err != nil {
+		s.Error(err)
+	}
+}
+
+// handles /shutdown
+func (s *shutdownmanagerContext) shutdownHandler(w http.ResponseWriter, r *http.Request) {
+	prometheusURL := fmt.Sprintf("http://%s:%d%s", s.envoyHost, s.envoyPort, s.prometheusPath)
```

Comment:
Do you think it might be worthwhile having this write something back to the connection so you can watch it with curl maybe?

Reply:
It writes to the logger later on each time it polls outputting the current connections and the min configured:

```
INFO[0005] min number of open connections found, shutting down...  context=shutdown-manager min_connections=0 open_connections=0
```

Reply:
I meant to the actual HTTP response, but it's probably not a good idea to be curling that endpoint anyway, so it's okay.

Reply:
To watch it with curl, the shutdown process would need to be separated from the HTTP handler, so that multiple requests could observe it. That separation would also make it robust against the shutdown handler accidentally being called multiple times, but would be more complex.

Reply:
I don't see a good use-case to watch with curl. To do that, you'd have to have access to the pod via kubectl since it's not exposed from the pod (i.e port-forward first, then curl). 

I would think you would tail the logs of shutdown-manager to watch what's happening.

Reply:
Yes, that's what I was trying to say earlier, this is fine for now.